File: src/game/rules.js
// Game rules: win detection, draw detection, valid moves
import { ROWS, COLS, WIN, EMPTY } from './constants.js';

/**
 * Returns true if `token` has WIN consecutive tokens anywhere on the board.
 * Checks horizontal, vertical, and both diagonals.
 */
export function checkWin(board, token) {
  // Horizontal
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c <= COLS - WIN; c++) {
      let found = true;
      for (let i = 0; i < WIN; i++) {
        if (board[r][c + i] !== token) { found = false; break; }
      }
      if (found) return true;
    }
  }

  // Vertical
  for (let c = 0; c < COLS; c++) {
    for (let r = 0; r <= ROWS - WIN; r++) {
      let found = true;
      for (let i = 0; i < WIN; i++) {
        if (board[r + i][c] !== token) { found = false; break; }
      }
      if (found) return true;
    }
  }

  // Diagonal down-right (\)
  for (let r = 0; r <= ROWS - WIN; r++) {
    for (let c = 0; c <= COLS - WIN; c++) {
      let found = true;
      for (let i = 0; i < WIN; i++) {
        if (board[r + i][c + i] !== token) { found = false; break; }
      }
      if (found) return true;
    }
  }

  // Diagonal up-right (/)
  for (let r = WIN - 1; r < ROWS; r++) {
    for (let c = 0; c <= COLS - WIN; c++) {
      let found = true;
      for (let i = 0; i < WIN; i++) {
        if (board[r - i][c + i] !== token) { found = false; break; }
      }
      if (found) return true;
    }
  }

  return false;
}

/** Returns true if the board is completely full (draw). */
export function isFull(board) {
  return board[0].every(cell => cell !== EMPTY);
}

/** Returns an array of column indices that are not yet full. */
export function validCols(board) {
  const cols = [];
  for (let c = 0; c < COLS; c++) {
    if (board[0][c] === EMPTY) cols.push(c);
  }
  return cols;
}

File: src/game/rules.py
# Game rules: win detection, draw detection, valid moves
from .constants import ROWS, COLS, WIN, EMPTY


def check_win(board, token):
    """
    Return True if `token` has WIN consecutive tokens anywhere on the board.
    Checks horizontal, vertical, and both diagonals.
    """
    # Horizontal
    for r in range(ROWS):
        for c in range(COLS - WIN + 1):
            if all(board[r][c + i] == token for i in range(WIN)):
                return True

    # Vertical
    for c in range(COLS):
        for r in range(ROWS - WIN + 1):
            if all(board[r + i][c] == token for i in range(WIN)):
                return True

    # Diagonal down-right (\)
    for r in range(ROWS - WIN + 1):
        for c in range(COLS - WIN + 1):
            if all(board[r + i][c + i] == token for i in range(WIN)):
                return True

    # Diagonal up-right (/)
    for r in range(WIN - 1, ROWS):
        for c in range(COLS - WIN + 1):
            if all(board[r - i][c + i] == token for i in range(WIN)):
                return True

    return False


def is_full(board):
    """Return True if the board is completely full (draw)."""
    return all(cell != EMPTY for cell in board[0])


def valid_cols(board):
    """Return a list of column indices that are not yet full."""
    return [c for c in range(COLS) if board[0][c] == EMPTY]
